text/0000-name-resolution.md

```diff
@@ -0,0 +1,420 @@
+item. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub use foo::Qux;
+}
+
+mod baz {
+    use foo::*;
+    use bar::*;
+
+    fn f(q: Qux) {}
+}
+```
+
+### non-public imports
+
+Currently `use` and `pub use` items are treated differently. Non-public imports
+will be treated in the same way as public imports, so they may be referenced
+from modules which have access to them. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    use foo::Qux;
+
+    mod baz {
+        use bar::Qux; // Ok
+    }
+}
+```
+
+
+### Glob imports of accessible but not public names
+
+Glob imports will import all accessible names, not just public ones. E.g.,
+
+```
+struct Qux;
+
+mod foo {
+    use super::*;
+
+    fn f(q: Qux) {} // Ok
+}
+```
+
+This change is backwards incompatible. However, the second rule above should
+address most cases, e.g.,
+
+```
+struct Qux;
+
+mod foo {
+    use super::*;
+    use super::Qux; // Legal due to the second rule above.
+
+    fn f(q: Qux) {} // Ok
+}
+```
+
+The below rule (though more controversial) should make this change entirely
+backwards compatible.
+
+Note that in combination with the above rule, this means non-public imports are
+imported by globs where they are private but accessible.
+
+
+### Globs and explicit names
+
+An explicit name may shadow a glob imported name without causing a name
+resolution error. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub struct Qux;
+}
+
+mod baz {
+    use foo::*;
+
+    struct Qux; // Shadows foo::Qux.
+}
+
+mod boz {
+    use foo::*;
+    use bar::Qux; // Shadows foo::Qux; note, ordering is not important.
+}
+```
+
+Note that shadowing is namespace specific. I believe this is consistent with our
+general approach to name spaces. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub trait Qux;
+}
+
+mod boz {
+    use foo::*;
+    use bar::Qux; // Shadows only in the type name space.
+
+    fn f(x: &Qux) {   // bound to bar::Qux.
+        let _ = Qux;  // bound to foo::Qux.
+    }
+}
+```
+
+This change is discussed in [issue 31337](https://github.com/rust-lang/rust/issues/31337).
+
+
+## Changes to the implementation
+
+Note: below I talk about "the binding table", this is sort of hand-waving. I'm
+envisaging a sets-of-scopes system where there is effectively a single, global
+binding table. However, the details of that are beyond the scope of this RFC.
+One can imagine "the binding table" means one binding table per scope, as in the
+current system.
+
+Currently, parsing and macro expansion happen in the same phase. With this
+proposal, we add import resolution to that mix too. Binding tables as well as
+the AST will be produced by libsyntax. Name lookup will continue to be done
+where name resolution currently takes place.
+
+To resolve imports, the algorithm proceeds as follows: we start by parsing as
+much of the program as we can; like today we don't parse macros. When we find
```

Comment:
What exactly does this mean? Don't parse macro definitions? invocations? the result of macro expansion?